README.md

About
=====

This package provides a redis-based token driver for
[Openstack Keystone](http://keystone.openstack.org/). This
works for the "Folsom" release and above.

### Known issues

* The optional tenant ID argument to `list_tokens()` is unused. No plans exist
  to change this, as it is not currently implementable without a full table scan
  of each and every token.

Installation
============

As the first step, be sure to activate the Python virtualenv environment
containing your Openstack Keystone setup, if applicable.

Install keystone-redis and dependencies:

    python setup.py install

To run the tests, make sure the `tests/` directory of the Openstack Keystone
repository is in your Python path:

    nosetests

Configuration
=============

To use keystone-redis in a Keystone configuration, edit the `driver` option
in the `token` section of `keystone.conf`.

    [token]
    driver = keystoneredis.token.Token

If the redis instance is only running on the local machine, you are all set.
To point it to a redis instance on another machine:

    [redis]
    connection = redis-instance.example.com
    # This option can be suffixed with a port number, e.g. :6379.

If you have read-only slaves that should handle any read operations, you can
configure a custom read-only connection:

    [redis]
    read_connection = redis-slave.example.com

To configure the redis database number (see the
[SELECT](http://redis.io/commands/select) command), use the `database` option in
the `redis` section. To configure the idle timeout, use the `idle_timeout`
option.

    [redis]
    idle_timeout = 200
    database = 0

Key Schema
==========

The keys written to redis are as follows

### Tokens

The token entries themselves will look like:

    token e3f75b94322746989868f07d61370086

### User Tokens List

To allow searching for all a particular user's tokens, a second key is generated
for every token in the following format:

    usertoken dXNlcmlk e3f75b94322746989868f07d61370086

The second piece of the key is a base64 encoded version of the user ID. In this
example, the user ID is `userid`.



debian/changelog
keystone-redis (0.1.1.wmf1) precise-wikimedia; urgency=low

  * source package automatically created by stdeb 0.6.0+git

 -- Ryan Lane <rlane@wikimedia.org>  Mon, 24 Feb 2014 17:44:26 +0000


tests/test_backend_redis.py
# vim: tabstop=4 shiftwidth=4 softtabstop=4

# Copyright 2012 OpenStack LLC
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.

import nose

from keystone import config
from keystone import test
from keystoneredis import token as token_redis
from keystoneredis.common import keys as redis_keys

import test_backend


CONF = config.CONF


class RedisToken(test.TestCase, test_backend.TokenTests):
    def setUp(self):
        super(RedisToken, self).setUp()
        self.config([test.etcdir('keystone.conf.sample'),
                     test.testsdir('test_overrides.conf'),
                     test.testsdir('backend_redis.conf')])
        self.token_api = token_redis.Token()
        self.token_api.flush_all()

    def tearDown(self):
        self.token_api.flush_all()
        super(RedisToken, self).tearDown()

    def test_token_list(self):
        tokens = self.token_api.list_tokens('testuserid')
        self.assertEquals(len(tokens), 0)
        token_id1 = self.create_token_sample_data()
        tokens = self.token_api.list_tokens('testuserid')
        self.assertEquals(len(tokens), 1)
        self.assertIn(token_id1, tokens)
        token_id2 = self.create_token_sample_data()
        tokens = self.token_api.list_tokens('testuserid')
        self.assertEquals(len(tokens), 2)
        self.assertIn(token_id2, tokens)
        self.assertIn(token_id1, tokens)
        self.token_api.delete_token(token_id1)
        tokens = self.token_api.list_tokens('testuserid')
        self.assertIn(token_id2, tokens)
        self.assertNotIn(token_id1, tokens)
        self.token_api.delete_token(token_id2)
        tokens = self.token_api.list_tokens('testuserid')
        self.assertNotIn(token_id2, tokens)
        self.assertNotIn(token_id1, tokens)

        # XXX: Please note: keystoneredis does NOT support tenant-specific
        #      tokens, so that piece of this test (originally copied from
        #      Keystone proper) has been removed.


keystoneredis/token.py
# vim: tabstop=4 shiftwidth=4 softtabstop=4

# Copyright 2012 Ian Good
# Copyright 2014 Ryan Lane
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.

"""Redis backends for the various services."""

import dateutil.parser
import copy

from keystone import exception
from keystone.openstack.common import timeutils
from keystone import token
from keystone.openstack.common import jsonutils

from common.session import RedisSession
from common import keys

class Token(RedisSession, token.Driver):

    def __init__(self, *args, **kwargs):
        RedisSession.__init__(self, *args, **kwargs)

    def flush_all(self):
        self.conn.flushall()

    def get_token(self, token_id):
        token_key = keys.token(token_id)
        value = self.readonly.get(token_key)
        if value:
            token = jsonutils.loads(value)
            if token.get('expires', None) is not None:
                token['expires'] = dateutil.parser.parse(token['expires'])
                if token['expires'] > timeutils.utcnow():
                    return token
            else:
                return token
        raise exception.TokenNotFound(token_id=token_id)

    def _set_keys(self, user_id, token_id, json_data, ttl_seconds):
        pipe = self.conn.pipeline()
        token_key = keys.token(token_id)
        if user_id:
            user_key = keys.usertoken(user_id['id'], token_id)
        if ttl_seconds is None:
            pipe.set(token_key, json_data)
            if user_id:
                pipe.set(user_key, '')
        else:
            pipe.setex(token_key, json_data, ttl_seconds)
            if user_id:
                pipe.setex(user_key, '', ttl_seconds)
        pipe.execute()

    def create_token(self, token_id, data):
        data_copy = copy.deepcopy(data)
        user_id = data_copy.get('user', None)
        if 'expires' not in data_copy:
            data_copy['expires'] = self._get_default_expire_time()
        json_data = jsonutils.dumps(data_copy)
        self._set_keys(user_id, token_id, json_data, self.ttl_seconds)
        return data_copy

    def _delete_keys(self, user_id, token_id):
        pipe = self.conn.pipeline()
        token_key = keys.token(token_id)
        pipe.delete(token_key, )
        if user_id is not None:
            user_key = keys.usertoken(user_id['id'], token_id)
            pipe.delete(user_key, )
        pipe.sadd(keys.revoked(), token_id)
        return pipe.execute()[0]

    def delete_token(self, token_id):
        data = self.get_token(token_id)
        user_id = data.get('user', None)
        if not self._delete_keys(user_id, token_id):
            raise exception.TokenNotFound(token_id=token_id)

    def list_tokens(self, user_id, tenant=None):
        pattern = keys.usertoken(user_id, '*')
        user_keys = self.readonly.keys(pattern)
        return [keys.parse_usertoken(key)[1] for key in user_keys]

    def list_revoked_tokens(self):
        return [{'id': s} for s in self.readonly.smembers(keys.revoked())]


keystoneredis/common/session.py
# vim: tabstop=4 shiftwidth=4 softtabstop=4

# Copyright 2012 Ian Good
# Copyright 2014 Ryan Lane
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.

"""Redis backends for the various services."""

import redis
from keystoneredis.common.redissl import Connection as SslConnection

from keystone import config
try:
    from oslo.config import cfg
except ImportError:
    from keystone.openstack.common import cfg

CONF = config.CONF

CONF.register_opt(cfg.StrOpt('read_connection', default=None), group='redis')
CONF.register_opt(cfg.StrOpt('connection', default='localhost'), group='redis')
CONF.register_opt(cfg.IntOpt('idle_timeout', default=200), group='redis')
CONF.register_opt(cfg.IntOpt('database', default=0), group='redis')
CONF.register_opt(cfg.StrOpt('password', default=''), group='redis')
CONF.register_opt(cfg.BoolOpt('ssl', default=False), group='redis')
CONF.register_opt(cfg.StrOpt('keyfile', default=None), group='redis')
CONF.register_opt(cfg.StrOpt('certfile', default=None), group='redis')
CONF.register_opt(cfg.StrOpt('ca_certs', default=None), group='redis')


class RedisSession(object):

    def __init__(self, **kwargs):
        read = kwargs.get('read_connection', CONF.redis.read_connection)
        local = kwargs.get('connection', CONF.redis.connection)
        database = kwargs.get('database', CONF.redis.database)
        password = kwargs.get('password', CONF.redis.password)
        idle_timeout = kwargs.get('idle_timeout', CONF.redis.idle_timeout)

        self.ttl_seconds = CONF.token.expiration
        self.local_client = self._create_client(local, database, password,
                                                idle_timeout)

        self.conn = self.local_client

        if read:
            self.readonly = self._create_client(read, database, password,
                                                idle_timeout)
        else:
            self.readonly = self.local_client

    def _create_client(self, connection, database, password, idle_timeout):
        try:
            host, port = connection.rsplit(':', 1)
        except ValueError:
            host = connection
            port = 6379

        if CONF.redis.ssl:
            pool = redis.ConnectionPool(connection_class=SslConnection,
                                        host=host, port=port, db=database,
                                        password=password,
                                        socket_timeout=idle_timeout,
                                        keyfile=CONF.redis.keyfile,
                                        certfile=CONF.redis.certfile,
                                        ca_certs=CONF.redis.ca_certs)
        else:
            pool = redis.ConnectionPool(host=host, port=port, db=database,
                                        password=password,
                                        socket_timeout=idle_timeout)

        return redis.Redis(connection_pool=pool)
